crates/build/src/solidity_metadata/natspec.rs

```diff
@@ -0,0 +1,262 @@
+// Copyright (C) ink! contributors.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use anyhow::Result;
+use contract_metadata::Contract;
+use ink_metadata::{
+    EventSpec,
+    InkProject,
+    MessageSpec,
+};
+use itertools::Itertools;
+use scale_info::{
+    form::PortableForm,
+    PortableRegistry,
+};
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+use super::abi;
+
+/// NatSpec developer documentation of the contract.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct DevDoc {
+    /// The version of the NatSpec format.
+    version: u8,
+    /// Kind of NatSpec documentation (i.e. "dev").
+    kind: NatSpecKind,
+    /// Author of the contract.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    author: Option<String>,
+    /// Describes the contract/interface.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    title: Option<String>,
+    /// Extra details for developers.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    details: Option<String>,
+    /// Storage developer documentation, keys are storage keys.
+    #[serde(rename = "stateVariables")]
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    state_variables: HashMap<String, ItemDevDoc>,
+    /// Function developer documentation, keys are canonical function signatures.
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    methods: HashMap<String, ItemDevDoc>,
+    /// Events developer documentation, keys are canonical event signatures.
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    events: HashMap<String, ItemDevDoc>,
+    /// Errors developer documentation, keys are canonical error signatures.
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    errors: HashMap<String, ItemDevDoc>,
+}
+
+/// NatSpec user documentation of the contract.
```

Comment:
Can you explain in a comment somewhere how developers can influence which text lands in `NatSpec` and which in `DevDoc`?

Reply:
So for Solidity, NatSpec (https://docs.soliditylang.org/en/latest/natspec-format.html) has a concept of "tags" (e.g. `@dev` and `@notice`) that represent the target audience (e.g. developer or user) for specific lines in docs e.g.
``` Solidity
/// @notice An NFT
/// @dev Inherits OpenZepplin ERC721 implementation
```

ink!/Rust don't really have a "native" equivalent to this, and I'm not even sure we need one because explorers seem to show both AFAICT. So I think we just treat ink!/Rust comments as `DevDoc` for now, and ignore (leave empty) the `UserDoc` fields.